### ISSUE TITLE
Separating cells of SRS

### ISSUE THREAD

**Issue by tingyuw** [JacquesCarette/Drasil#2346]
How the document is printed currently is the printer calls printLO and maps the contents together. 
![image](https://user-images.githubusercontent.com/47548045/116967542-83164c00-ac80-11eb-8ced-d78e1a5aa936.png)

The HDiv maps the printLO layout objects to a huge section. 
![image](https://user-images.githubusercontent.com/47548045/116967735-00da5780-ac81-11eb-8ec5-868b57cd6868.png)
![image](https://user-images.githubusercontent.com/47548045/116975881-5406d700-ac8e-11eb-928c-f4ecc4bbe22c.png)

Thus, in the HTML, subsections are wrapped inside the ‘main’ sections, and each main section consisted of several subsections. If we separate the cells by the main sections, it would still be too large to maintain.
![image](https://user-images.githubusercontent.com/47548045/116974496-39336300-ac8c-11eb-9f95-0a5e5cb0053a.png)

So I was trying to separate the contents by each subsection, which means I need to ‘cut’ the printLO in the right place (maybe at the end of each subsection). That is, HDiv should not map the printLO, at least not at the time when we are still trying to print more information in the documents (the information for building a cell). The problem is how can I cut the printLO? I would like to print the document well without mapping the contents.

Another thing is, how do we know where is a section (or cell, if we would like to separate cells by sections)? This can depend on how we can cut the printLO (if applicable). For the SRS, if we can cut the printLO at any place we want (at the end of each data object for example), maybe we can use the headers as the separation standards at this point, or follow how the HTML handle subsections. And if we can cut the printLO properly, we might not need to expand the LayoutObj.
![image](https://user-images.githubusercontent.com/47548045/116974572-549e6e00-ac8c-11eb-8c90-adb3b44eabcd.png)

For the future application (such as the projectile example), to know what is a cell, my thought is to distinguish the headers and consider data objects. Some data objects could be a cell itself, such as tables, lists, and figures, and the other data objects can be separated by headers. Since it's hard to define what is a cell, we could figure out a pattern we want to separate the contents.

**Comment by JacquesCarette** [comment]
You are explaining everything to us in terms of "here is what the code does now" and "here are the changes I would like to make to the code".  You have also only shown us how it is done in the HTML printer, but not the latex printer. There could be large differences.

Note that to the question about what to do with the code: you most definitely should be thinking in terms of a completely new printer for generating notebooks. In there, you'll have complete control of the logic. Nevertheless, I don't think that will be enough.

You still have not addressed the following crucial questions:
- what seems to be represented, conceptually, by an `HDiv`?
- what is the logic to generate an `HDiv` from "higher up"? [i.e. where do `HDiv` come from?]
- what is, conceptually, a 'cell'? [I know this is hard, and you try to avoid this in your last paragraph above]
- correspondingly, from a higher-level representation of a document (i.e. the things that get mapped to `LayoutObj`), what is the information that should indicate that new cells are needed?

What we need is a "working definition" of a cell, even if it's wrong. It will be a starting point. For an extreme example, we spent a whole summer on "what is a reference", and had a dozen or so attempts at it. Our current definition is still unsatisfactory! But it does work, so we use it until we're forced to change it again.

**Comment by tingyuw** [comment]
The translation from contents (Drasil original data types) to printing layoutobj is handled in ``Import.hs``. Unlike other data types (e.g. Table or Paragraph), ``HDiv`` is not a data type we have in Drasil original language; it is specifically for printing and is not a one-to-one transformation. ``HDiv`` is called in the ``sec`` to concatenate each section. I still need to trace back the code more to fully understand how it works though. 
![image](https://user-images.githubusercontent.com/47548045/117173813-2491c080-ad9b-11eb-86c3-5bb8eee48dd5.png)

The latex printer prints in a similar way HTML do but without the wrapping. And right, it's not about how to print the document, but how to handle the contents. So it brings back to the question of what is a cell. I think a cell is a small paragraph which concludes sufficient information (readers can take away something in a single cell) but still not too much. For example, a code cell can do an easy calculation and output a value and a markdown cell can be a table with a caption. On the practical side, my current thought would be using the headers as an indication for a new cell. A header usually indicates a new paragraph which contains new information. Each header and the contents under that header is in a cell, and we add a new cell when the next header comes up.

**Comment by JacquesCarette** [comment]
Thank you.  Yes, defining a cell as a kind of "unit of content" that is kind of a "leaf" in a structured document, makes sense. It's not the lowest level of 'content', but somehow the lowest leave of "display content" that tells a coherent piece of information.

My feeling is that we need to somehow create an indicator for such content units. It might be that in the `Import.hs`, cell information is generated into the "layout" document, but that this information might be ignored by both the latex and HTML printers, but used by the Markdown / notebook printer.

**Comment by smiths** [comment]
I won't pretend to understand all the details, but I like the direction this conversation is going.  A cell as a king of "unit of content" sounds right to me.  Should the cell also know what kind of content it is?  For Jupyter notebooks the content can either be markdown or code.  Would it make sense for us to breakdown the markdown category further, maybe into equations, text, tables and figures?

**Comment by JacquesCarette** [comment]
Yes, knowing what kind of cell it is makes sense.

The important thing is the granularity. For example, there could be multiple text cells one after another, because each is a kind of self-contained paragraph.

**Comment by tingyuw** [comment]
Yes, I think the changes will be made in the ``Import.hs``. I was thinking of handling the contents in cells instead of sections (i.e. change how the ``HDiv`` maps the contents) for the notebook printer, and the cell information can be added when printing. I would like to change the ``HDiv`` because the document format isn't the same as the SRS in the future application. The sections are not decided in the way they are now, and the contents won't be added in the same way. We might need new languages and a new way of handling the contents of the notebook in the future. And for whether the content is markdown or code, I think this can be distinguished when adding the information in Drasil.

**Comment by tingyuw** [comment]
That makes sense. How about first trying to add the contents of the projectile example in Drasil and create a proper language for inputting notebooks then decide the best layout for it? Since how we input the notebook contents might change the way we handle them in ``Import``. And we can comeback to the SRS in Jupyter after.

**Comment by JacquesCarette** [comment]
Yes, that sounds like a good plan. A "language of notebooks" will be somewhat different than the one for SRS. The commonalities will be interesting too!

**Comment by tingyuw** [comment]
Great. I'm looking at the SRS language these days, and I will create new issues regarding the structure of the notebook and our expectations for it in a few days.

**Comment by tingyuw** [comment]
Update: Currently, the cells are separated by main sections, both work fine in SRS and projectile motion. The document can be checked by running the notebook branch, under /Projectile/Jupyter. (Ignore some of the references and equations don't print well at this moment, I'm still editing the code constantly.) I introduced a new ``LayoutObj``-- ``Cell``, and I change the outermost layer of ``HDiv``(of each section) to ``Cell``. However, I think each cell block is still too big to maintain. 

How the current printer translates the contents into ``LayoutObj`` makes it hard to separate cells in the appropriate depth. We don't need to wrap the contents in a bigger content block in notebooks, we should print it linearly, especially when we want to add code cells in between. So I was thinking to change how the ``HDiv`` works in notebooks, for example, there is no need to wrap ``Eqnblock`` in ``HDiv`` in notebooks, and we also might not need to map the contents together, just print one content block at a time.

**Comment by JacquesCarette** [comment]
Great update. But I would really like to see a design of this part, not just an implementation. I don't want to put a lot of 'smarts' in the printers, they should be very straightforward code. That means that I want to put the smarts into the source language that the printers translate (and perhaps a bit into `LayoutObj` too, but as a second choice).

**Comment by tingyuw** [comment]
I am having trouble to 'cut' the contents in the right place. What I'm thinking is to make each subsection a ``Cell LayoutObj``, and combine them to ``[LayoutObj]`` for printing but it doesn't work well. Even though I can make each subsection a cell, it is still 'inside' a bigger content. I would like to cut the contents whenever it meets a header, for example, the main header and it's overview paragraph is a ``Cell`` and the following subheader and its contents should be in another ``Cell``. Is it possible to do this in the current ``Import`` or should I add new languages for cells in the source like the code block in ``RawContent``?

**Comment by balacij** [comment]
@tingyuw Can you please also check on the state of this ticket too?

EDIT: This was not fixed.

**Comment by balacij** [comment]
#5201 reminded me of this ticket because it helped me better understand how information really flows in the lesson plan generator:

https://github.com/JacquesCarette/Drasil/blob/3e5662cd690a8d9a859fb64bc1d87d779c097e26/code/drasil-lesson-plan/lib/Drasil/LessonPlan/Renderer.hs#L45-L66

Disclaimers:

* SDL/TDL do not currently exist. The names are just what I call them. (I dislike seeing 3 `Document` types. Referring to which `Document` should be unambiguous for us. I won't argue for or against them right now because I don't quite like them yet.)
* `genJupyterLessonPlan` is a rendering function from `drasil-printers` that does some special stuff that probably shouldn't appear at all. The magic that differentiates `genJupyterLessonPlan` from `genJupyterSRS` needs to be moved to a rendering option _between_ SDL and TDL.

Presently, the SRS is rendered as a single Markdown cell. One 'easy' and hopefully uncontroversial change is to chop up the cells according to the sections. Note that we are not nesting cells. We are creating a flat list of section breaking cells. These section breaking cells (that introduce section headers) should also contain the `x.0` content (i.e., the few preliminary paragraphs/figures/etc. content that appear before any subsections).

Where _could_ this be done? Between the conversion of SDL and TDL! The SDL contains this information:

https://github.com/JacquesCarette/Drasil/blob/98704433ce56657d8169c4c3e3f5876244cd9bb8/code/drasil-lang/lib/Language/Drasil/Document/Sections.hs#L28-L40

https://github.com/JacquesCarette/Drasil/blob/98704433ce56657d8169c4c3e3f5876244cd9bb8/code/drasil-lang/lib/Language/Drasil/Document/Sections.hs#L144-L147

Notes:

1. We should consider erasing `SecCons` in favour of having two lists (`prelimCont :: [Content]` and `subSecs :: [Section]`) in the `Section` encoding. `SecCons` does not appear to add enough value (see `rg "SecCons" -ths`).
2. We will most likely need to let _some_ 'grouping' knowledge leak through into TDL, but we should see this as a good thing. That grouping contains meta-information we can use here (in the Jupyter renderer) but also in other places (e.g., if we had an org-mode code generator or wanted to create HTML options to mask more things behind `<details>..</details>` summaries).

**Comment by balacij** [comment]
Also, I think we should refrain from using `HDiv` (from TDL) any further (and try to remove it instead). It looks like a leaky abstraction.

**Comment by JacquesCarette** [comment]
Should parts of the discussion we had yesterday modify this issue a bit?

For sure "grouping" has semantic content. Calling it `HDiv` in a high-level data-structure is a bad idea -- but using `div` in the generated html is a good idea.

**Comment by balacij** [comment]
Aside from terminology used to refer to the "transformations," I don't immediately see how it would. The suggestion I made above is fairly tame.

Agreed.
